### PR TITLE
Pseudocode for using IoT Hub Client

### PR DESCRIPTION
**Pull request by danewalton**

### CODE REVIEW

File: sdk/docs/iot/README.md [hunk 1]
```diff
@@ -174,6 +174,126 @@ void my_telemetry_func()
 }
 ```
 
+### IoT Hub Pseudocode Implementation
```
Comment (ericwol-msft) [inline sdk/docs/iot/README.md:177]: What makes this Pseudocode? Snippet?

File: sdk/docs/iot/README.md [hunk 2]
```diff
@@ -174,6 +174,126 @@ void my_telemetry_func()
...
+
+  //The message could be for three features so parse the topic to see which it is for
+  az_iot_hub_client_method_request method_request;
+  az_iot_hub_client_c2d_request c2d_request;
+  az_iot_hub_client_twin_response twin_response;
+  if(az_iot_hub_client_methods_parse_received_topic(&client, incoming_topic, &method_request) == AZ_OK)
+  {
+    //Handle the method request
+  }
+  else if(az_iot_hub_client_c2d_parse_received_topic(&client, incoming_topic, &c2d_request) == AZ_OK)
```
Comment (ericwol-msft) [inline sdk/docs/iot/README.md:285]: nit, space after **if**

File: sdk/docs/iot/README.md [hunk 3]
```diff
@@ -174,6 +174,126 @@ void my_telemetry_func()
...
+  {
+    //Handle the c2d request
+  }
+  else if(az_iot_hub_client_twin_parse_received_topic(&client, incoming_topic, &twin_response) == AZ_OK)
```
Comment (ericwol-msft) [inline sdk/docs/iot/README.md:289]: space after if

File: sdk/docs/iot/README.md (hunk 1 above)
Comment (ericwol-msft) [reply sdk/docs/iot/README.md:177]: u call it snippet int the text

File: sdk/docs/iot/README.md [hunk 4]
```diff
@@ -174,6 +174,126 @@ void my_telemetry_func()
...
+```C
+#include <az/core/az_result.h>
+#include <az/core/az_span.h>
+#include <az/iot/az_iot_hub_client.h>
+
+az_iot_hub_client my_client;
+static az_span my_iothub_hostname = AZ_SPAN_LITERAL_FROM_STR("<your hub fqdn here>");
+static az_span my_device_id = AZ_SPAN_LITERAL_FROM_STR("<your device id here>");
+
+//Make sure to size the buffer to fit the user name (100 is an example)
```
Comment (ericwol-msft) [inline sdk/docs/iot/README.md:190]:
```suggestion
//Make the buffer is big enough to fit the user name (100 is an example)

```

File: sdk/docs/iot/README.md [hunk 5]
```diff
@@ -174,6 +174,126 @@ void my_telemetry_func()
...
+static char my_mqtt_user_name[100];
+static size_t my_mqtt_user_name_length;
+
+//Make sure to size the buffer to fit the client id (16 is an example)
```
Comment (ericwol-msft) [inline sdk/docs/iot/README.md:194]:
```suggestion
//Make sure the buffer is big enough to fit the client id (16 is an example)

```

File: sdk/docs/iot/README.md [hunk 6]
```diff
@@ -174,6 +174,126 @@ void my_telemetry_func()
...
+#include <az/core/az_span.h>
+#include <az/iot/az_iot_hub_client.h>
+
+az_iot_hub_client my_client;
+static az_span my_iothub_hostname = AZ_SPAN_LITERAL_FROM_STR("<your hub fqdn here>");
+static az_span my_device_id = AZ_SPAN_LITERAL_FROM_STR("<your device id here>");
+
+//Make sure to size the buffer to fit the user name (100 is an example)
+static char my_mqtt_user_name[100];
+static size_t my_mqtt_user_name_length;
```
Comment (ericwol-msft) [inline sdk/docs/iot/README.md:192]: not used in the snippet

File: sdk/docs/iot/README.md [hunk 7]
```diff
@@ -174,6 +174,126 @@ void my_telemetry_func()
...
+
+static char telemetry_topic[128];
+static char telemetry_payload[] = "Hello World";
+
+void handle_iot_message(mqtt_client_message* msg);
+
+int main()
+{
+  //Get the default IoT Hub options
+  az_iot_hub_client_options options = az_iot_hub_client_options_default();
```
Comment (ericwol-msft) [inline sdk/docs/iot/README.md:208]: this is an optional param.. leave it out of the snippet?

File: sdk/docs/iot/README.md [hunk 8]
```diff
@@ -174,6 +174,126 @@ void my_telemetry_func()
...
+  //Initialize the client with hostname, device id, and options
+  az_iot_hub_client_init(&my_client, my_iothub_hostname, my_device_id, &options);
+
+  //Get the MQTT user name to connect
+  az_iot_hub_client_get_user_name(&my_client, my_mqtt_user_name, 
+                sizeof(my_mqtt_user_name), NULL);
+
+  //Get the MQTT client id to connect
+  az_iot_hub_client_get_client_id(&my_client, my_mqtt_client_id, 
+                sizeof(my_mqtt_client_id), &my_mqtt_client_id_length);
```
Comment (ericwol-msft) [inline sdk/docs/iot/README.md:219]: my_mqtt_client_id_length not defined..needed?

File: sdk/docs/iot/README.md [hunk 9]
```diff
@@ -174,6 +174,126 @@ void my_telemetry_func()
...
+
+
+  //Initialize chosen MQTT client with necessary parameters (example params shown)
```
Comment (ericwol-msft) [inline sdk/docs/iot/README.md:222]:
```suggestion
  //Initialize MQTT client with necessary parameters (example params shown)

```

File: sdk/docs/iot/README.md [hunk 10]
```diff
@@ -174,6 +174,126 @@ void my_telemetry_func()
...
+
+  //Get the MQTT user name to connect
+  az_iot_hub_client_get_user_name(&my_client, my_mqtt_user_name, 
+                sizeof(my_mqtt_user_name), NULL);
+
+  //Get the MQTT client id to connect
+  az_iot_hub_client_get_client_id(&my_client, my_mqtt_client_id, 
+                sizeof(my_mqtt_client_id), &my_mqtt_client_id_length);
+
+
```
Comment (ericwol-msft) [inline sdk/docs/iot/README.md:221]: extra space

File: sdk/docs/iot/README.md [hunk 11]
```diff
@@ -174,6 +174,126 @@ void my_telemetry_func()
...
+  //Subscribe to twin response topic 
+  mqtt_client_subscribe(&my_mqtt_client, AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_SUBSCRIBE_TOPIC);
+
+  //Connect to the IoT Hub with your chosen mqtt stack
+  mqtt_client_connect(&my_mqtt_client, my_mqtt_user_name, my_device_cert);
+
+  //This example would run to receive any incoming message and send a telemetry message five times
+  int iterations = 0;
+  mqtt_client_message msg;
+  while(iterations < 5)
```
Comment (ericwol-msft) [inline sdk/docs/iot/README.md:244]:
```suggestion
  while(iterations++ < 5)

```

File: sdk/docs/iot/README.md [hunk 12]
```diff
@@ -174,6 +174,126 @@ void my_telemetry_func()
...
+  {
+    if(mqtt_client_receive(&msg))
+    {
+      handle_iot_message(&msg);
+    }
+
+    send_telemetry_message();
+
+    iterations++;
```
Comment (ericwol-msft) [inline sdk/docs/iot/README.md:253]: move into while

File: sdk/docs/iot/README.md (hunk 1 above)
Comment (danewalton) [reply sdk/docs/iot/README.md:177]: Yea that's fine wasn't exactly sure what to call it

File: sdk/docs/iot/README.md [hunk 13]
```diff
@@ -174,6 +174,126 @@ void my_telemetry_func()
...
+  az_iot_hub_client_method_request method_request;
+  az_iot_hub_client_c2d_request c2d_request;
+  az_iot_hub_client_twin_response twin_response;
+  if(az_iot_hub_client_methods_parse_received_topic(&client, incoming_topic, &method_request) == AZ_OK)
+  {
+    //Handle the method request
+  }
+  else if(az_iot_hub_client_c2d_parse_received_topic(&client, incoming_topic, &c2d_request) == AZ_OK)
+  {
+    //Handle the c2d request
```
Comment (ericwol-msft) [inline sdk/docs/iot/README.md:287]:
```suggestion
    //Handle the c2d message

```

File: sdk/docs/iot/README.md (hunk 7 above)
Comment (danewalton) [reply sdk/docs/iot/README.md:208]: Sure I can. The intention was to draw attention to what the last param is. If I just pass in `NULL`, people won't exactly know what it is for.

File: sdk/docs/iot/README.md (hunk 8 above)
Comment (danewalton) [reply sdk/docs/iot/README.md:219]: changed to NULL

File: sdk/docs/iot/README.md (hunk 7 above)
Comment (danewalton) [reply sdk/docs/iot/README.md:208]: Actually unless you're really opposed, I'll probably keep in for that reason

**Pull request merged**